shutdown.go

```diff
@@ -0,0 +1,67 @@
+package gokit
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+// Server will create a http.Server from the Go standard library
+type Server struct {
+	Addr    string        // TCP address to listen on, ":http" if empty
+	Handler http.Handler  // handler to invoke, http.DefaultServeMux if nil
+	Timeout time.Duration // Timeout to wait for shutdown
+}
+
+// ListenAndServe mirrors the function from the Go standard library
+// ListenAndServe always returns a non-nil error
+func (s *Server) ListenAndServe() error {
+	server := http.Server{
+		Addr:    s.Addr,
+		Handler: s.Handler,
+	}
+
+	// We want to report the listener is closed.
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	done := make(chan struct{})
+	var output error
+	// Start the listener.
+	go func() {
+		output = server.ListenAndServe()
+		close(done)
+		wg.Done()
+	}()
+
+	// Listen for an interrupt signal from the OS. Use a buffered
+	// channel because of how the signal package is implemented.
+	osSignals := make(chan os.Signal, 1)
+	signal.Notify(osSignals, os.Interrupt, syscall.SIGTERM)
+
+	// Wait for a signal to shutdown.
+	select {
+	case <-done:
+	case <-osSignals:
+		ctx, cancel := context.WithTimeout(context.Background(), s.Timeout)
+		defer cancel()
+
+		// Attempt the graceful shutdown by closing the listener and
+		// completing all inflight requests.
+		if err := server.Shutdown(ctx); err != nil {
+			output = err
```

Comment:
Will this lead to a race in whether the err is assigned here v.s. if it is assigned in line 36? What happens if you `Shutdown` a server that didn't start up?

Reply:
It's intentional to over write the error with this error, because it's should be a running server at this point in time and this `shutdown` error should be captured instead.

If a server didn't start up, it will
1. Close the `done` channel
1. Trigger the select on `<-done`
1. Not run this code

If the server did not start up and a `SIGINT` was triggered before shutdown AND causing a non-nil error, it seems reasonable to log the non-nil error instead. What do you think?

Reply:
Oh oops, I thought like a `switch` it will fallthrough. Not sure what I was thinking there.

Reply:
seems like fallthrough in Go is more explicit https://github.com/golang/go/wiki/Switch#fall-through